Assets/Mirror/Runtime/NetworkReaderExt.cs

```diff
@@ -0,0 +1,252 @@
+        public static Color ReadColor(this NetworkReader reader) => new Color(reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle());
+        public static Color32 ReadColor32(this NetworkReader reader) => new Color32(reader.ReadByte(), reader.ReadByte(), reader.ReadByte(), reader.ReadByte());
+        public static Quaternion ReadQuaternion(this NetworkReader reader) => new Quaternion(reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle());
+        public static Rect ReadRect(this NetworkReader reader) => new Rect(reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle());
+        public static Plane ReadPlane(this NetworkReader reader) => new Plane(reader.ReadVector3(), reader.ReadSingle());
+        public static Ray ReadRay(this NetworkReader reader) => new Ray(reader.ReadVector3(), reader.ReadVector3());
+
+        public static Matrix4x4 ReadMatrix4x4(this NetworkReader reader)
+        {
+            return new Matrix4x4
+            {
+                m00 = reader.ReadSingle(),
+                m01 = reader.ReadSingle(),
+                m02 = reader.ReadSingle(),
+                m03 = reader.ReadSingle(),
+                m10 = reader.ReadSingle(),
+                m11 = reader.ReadSingle(),
+                m12 = reader.ReadSingle(),
+                m13 = reader.ReadSingle(),
+                m20 = reader.ReadSingle(),
+                m21 = reader.ReadSingle(),
+                m22 = reader.ReadSingle(),
+                m23 = reader.ReadSingle(),
+                m30 = reader.ReadSingle(),
+                m31 = reader.ReadSingle(),
+                m32 = reader.ReadSingle(),
+                m33 = reader.ReadSingle()
+            };
+        }
+
+        public static byte[] ReadBytes(this NetworkReader reader, int count)
+        {
+            byte[] bytes = new byte[count];
+            reader.ReadBytes(bytes, count);
+            return bytes;
+        }
+
+        public static Guid ReadGuid(this NetworkReader reader) => new Guid(reader.ReadBytes(16));
+        public static Transform ReadTransform(this NetworkReader reader) => reader.ReadNetworkIdentity()?.transform;
+        public static GameObject ReadGameObject(this NetworkReader reader) => reader.ReadNetworkIdentity()?.gameObject;
+
+        public static NetworkIdentity ReadNetworkIdentity(this NetworkReader reader)
+        {
+            uint netId = reader.ReadPackedUInt32();
+            if (netId == 0) return null;
+
+            if (NetworkIdentity.spawned.TryGetValue(netId, out NetworkIdentity identity))
+            {
+                return identity;
+            }
+
+            if (LogFilter.Debug) Debug.Log("ReadNetworkIdentity netId:" + netId + " not found in spawned");
+            return null;
+        }
+    }
+}
```

Comment:
why not keep the extensions in NetworkReader/Writer like before?
how much extra magic would this require in the weaver? can't we just find all NetworkReader.X functions no matter if static or not?

Reply:
one problem is that we have multiple methods that would match.  For example,  
`ReadInt32`  and `ReadPackedInt32`  would both be found.   So I would have to make some special case for NetworkReader/NetworkWriter and then also a special case for ReadInt32 to ignore it.

I would rather mirror didn't have special cases.  I would rather NetworkWriter was kept minimal and then all the fancy writers were added as extensions.

Reply:
@paulpach well.. we could remove readpacked. we were never bandwidth limited. it would be less computations, way less code for something that is supposed to be simple, less prone to errors or users always using it and not knowing it it can also send _more_ data in many cases.

also, Weaver doesn't use WritePacked for _all_ int/short/byte types anymore with this PR, right? in that case, it's even better if bytes are really written as 1 byte, etc.

I never liked packed writing. it's 2019, people download 16GB movies. why should we constantly worry about 1-2 bytes here.

Reply:
and most importantly: Unity is single threaded. we can't afford unnecessary computations, especially for large MMOs etc.

Reply:
ok,  so basically you want me to have 2 mechanisms for serializers:  one is the methods in NetworkWriter and another one is extension methods.  I fail to see why we would want that extra complexity in the weaver that now has to look for two set of methods that are so different.

We can get rid of WritePacked,  but let's do that as a separate PR.   This PR is complicated enough already without changing serialization logic.

Reply:
Bandwidth usage cannot be ignored because those hosting the servers and paying per GB for game networking traffic are hyper-sensitive to bandwidth abuse.  They definitely need us to be as tight as practical, because managing hosting cost with more memory or CPU is a lot easier and likely cheaper than bandwidth.

Reply:
@MrGadget1024 of course, bandwidth still matters.
but here is the difference: if bandwidth becomes an issue, it means we need to pay more $ / month.
if single thread CPU speed becomes an issue, we do have a problem. those things only get so fast.

Reply:
@paulpach can we move writer/reader extensions into the original NetworkReader/Writer.cs and a little bit below in the file? having them in two separate ones is a bit strange I think.

And maybe rename to NetworkWriterExtensions instead of 'Ext' - this makes it easier for non english speaking people, if it's directly obvious what it is.

Reply:
I am confused you want me to rename NetworkWriterExt or get rid of it?

Reply:
I want to decouple this feature from packed data discussion. It is completely orthogonal.

Reply:
Please rename it to NetworkReaderExtension. humans are very good at reading words, but less good at reading abbreviations.

Reply:
Done

Reply:
Ohhhhhh,   I get what you are saying,  you want 2 classes in NetworkWriter.cs,  ok,  one sec...